topic-operator/src/main/java/io/strimzi/operator/topic/cruisecontrol/CruiseControlHandler.java

```diff
@@ -0,0 +1,205 @@
+     * @param cruiseControlClient Cruise Control client.
+     */
+    public CruiseControlHandler(TopicOperatorConfig config,
+                                TopicOperatorMetricsHolder metricsHolder,
+                                CruiseControlClient cruiseControlClient) {
+        this.config = config;
+        this.metricsHolder = metricsHolder;
+        this.cruiseControlClient = cruiseControlClient;
+    }
+
+    /**
+     * Send a topic_configuration request to create a task for replication factor change of one or more topics.
+     * This should be called when one or more .spec.replicas changes are detected.
+     * Note that this method also updates the KafkaTopic status.
+     * 
+     * @param reconcilableTopics Pending replicas changes.
+     * @return Results with status updates.
+     */
+    public Results requestPendingChanges(List<ReconcilableTopic> reconcilableTopics) {
+        var results = new Results();
+        if (reconcilableTopics.isEmpty()) {
+            return results;
+        }
+        results.addAll(updateToPending(reconcilableTopics, "Replicas change pending"));
+
+        var timerSample = TopicOperatorUtil.startExternalRequestTimer(metricsHolder, config.enableAdditionalMetrics());
+        try {
+            LOGGER.debugOp("Sending topic configuration request, topics {}", topicNames(reconcilableTopics));
+            var kafkaTopics = reconcilableTopics.stream().map(ReconcilableTopic::kt).collect(Collectors.toList());
+            var userTaskId = cruiseControlClient.topicConfiguration(kafkaTopics);
+            results.addAll(updateToOngoing(reconcilableTopics, "Replicas change ongoing", userTaskId));
+        } catch (Throwable t) {
+            results.addAll(updateToFailed(reconcilableTopics, format("Replicas change failed, %s", getRootCause(t).getMessage())));
+        }
+        TopicOperatorUtil.stopExternalRequestTimer(timerSample, metricsHolder::cruiseControlTopicConfig, config.enableAdditionalMetrics(), config.namespace());
+        
+        return results;
+    }
+
+    /**
+     * Send a user_tasks request to check the state of ongoing replication factor changes.
+     * This should be called periodically to update the active tasks cache and KafkaTopic status.
+     * Note that this method also updates the KafkaTopic status.
+     *
+     * @param reconcilableTopics Ongoing replicas changes.
+     * @return Results with status updates.
+     */
+    public Results requestOngoingChanges(List<ReconcilableTopic> reconcilableTopics) {
+        var results = new Results();
+        if (reconcilableTopics.isEmpty()) {
+            return results;
+        }
+        
+        var groupByUserTaskId = reconcilableTopics.stream()
+            .filter(rt -> hasReplicasChangeStatus(rt.kt()) && rt.kt().getStatus().getReplicasChange().getSessionId() != null)
+            .map(rt -> new ReconcilableTopic(new Reconciliation("", KafkaTopic.RESOURCE_KIND, "", ""), rt.kt(), rt.topicName()))
```

Comment:
Why are these empty strings correct?

Reply:
Unless there's a reason for them to be different why not just copy the values over? That reduces the possibility for an accident in the future, especially if this became a pattern used in several places in the operator: We then have Reconciliations-that-are-not-Reconciliations floating around waiting to trip us up.

Reply:
Effectively, there is no reason for this mapping. It must have been some test I did at the time. Removed.